Update Getting Start With Rails guide to account for Dev Containers (#51637)
rails/rails: guides/source/getting_started.md

After reading this guide, you will know:

* How create a new Rails application

> * How to work with your Rails application in a Dev Container

Reviewer: What about?

```suggestion
* How to create, set up, and run a Rails application (in a Dev Container)
```

`blog`, a (very) simple weblog. Before you can start building the application,
you need to make sure that you have Rails itself installed.

This guide helps you get set up with [Developer Containers (or Dev Containers for short)](https://containers.dev/)
for a full-featured development environment. Dev Containers are used to run your
Rails application in a container, without needing to install Rails or its dependencies
directly on your machine. This is the fastest way to get your Rails application up and running.

Alternatively, to install Rails directly on your machine, you can read through the

> [Installing Rails guide](installing-rails.html)

Reviewer 2: ```suggestion
[Installing Rails guide](installing_rails.html)
```

`blog`, a (very) simple weblog. Before you can start building the application,
you need to make sure that you have Rails itself installed.

This guide makes use of Dev Containers, which allow you to work with your Rails
application in a container, without needing to install Rails or its dependencies
directly on your machine. This is the fastest way to get started with Rails.

If you would like to install Rails directly on your machine, you can see the
[Installing Rails guide](installing-rails.html)

NOTE: The examples below use `$` to represent your terminal prompt in a UNIX-like OS,
though it may have been customized to appear differently. If you are using Windows,

> your prompt will look something like `C:\source_code>`.

Reviewer: With the re-arrangement of the sections, this note feels a little out of place before the installation section, maybe we could move it to after the intro paragraph in "Creating a Blog Application"?

> ### Installing rails-new

Reviewer: We have a sub-heading in this section also named `#### Installing rails-new` as well which feels a bit confusing. Maybe we could rename this main section to 

```suggestion
### Setup and Installation
```

What do you think?

This guide utilizes the `rails-new` tool to create a Rails app on your machine. We will
then run our application in a Dev Container using VSCode. Both rails-new and the
Dev Container run in a containerized environment using Docker. So we will need to install
VSCode and Docker as well as `rails-new`.

#### Installing Docker

Dev Containers are run using Docker. You can install Docker by following the
installation instructions for your operating system in the [Docker docs](https://docs.docker.com/desktop/).

Once Docker has been installed, open it from your application folder to begin running

> the Docker engine on your machine.

Reviewer: Maybe we can also add something like this to connect Docker and DevContainer: 

Dev Containers will leverage Docker's containers to allow you to encapsulate your development environment and dependencies within isolated containers.

#### Installing VSCode

> The easiest way of working with Dev Containers is using the VSCode editor. You can
> install VSCode by downloading it from [the website](https://code.visualstudio.com/).

Reviewer: Do you think we could add a sentence that describes how VS code is beneficial when working with Dev Containers. Maybe something like: 

```suggestion
VSCode's integration with Dev Containers helps to streamline your development workflow. It helps to quickly spin up containers, access terminal commands, debug code, and utilize extensions. You can install VSCode by downloading it from [the website](https://code.visualstudio.com/).
```

Reviewer: In addition, I thought we need to install the [Visual Code extension CLI](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) to start and attach the development container? Maybe that's not correct though?

If it turns out wew do need the extension then we coul summarise this for folks to explain its purpose "The Visual Studio Code Dev Containers extension allows you to open any folder inside (or mounted into) a container and take advantage of Visual Studio Code's full feature set. A [devcontainer.json file](https://code.visualstudio.com/docs/devcontainers/containers#_create-a-devcontainerjson-file) in your project tells VS Code how to access (or create) a development container with a well-defined tool and runtime stack. You can install it by..."

Author: Argh! Yes thank you I missed that step.

#### Installing rails-new

> `rails-new` is a tool which can generate a new Rails application without needing
> Ruby or Rails installed on your machine. To install rails-new, follow the installation
> instructions [in the readme](https://github.com/rails/rails-new?tab=readme-ov-file#installation).

Reviewer: Same as above, connecting up the installation tools could be useful?

```suggestion
`rails-new`  generates a new Rails application for you without having to install Ruby on your machine. It uses Docker to generate the Rails application, thus allowing Docker to take care of installing the correct Ruby and Rails versions for you.

To install rails-new, follow the installation instructions [in the readme](https://github.com/rails/rails-new?tab=readme-ov-file#installation).
```

> instructions [in the readme](https://github.com/rails/rails-new?tab=readme-ov-file#installation).

Reviewer: @andrewn617 I was just trying to install this locally and I don't see any installation instructions in the readme for `rails-new`?

Author: I was chatting with Rafael about this today and it's next on my todo list. I'll have a PR open for it in the next 24 hours.

|.gitignore|This file tells git which files (or patterns) it should ignore. See [GitHub - Ignoring files](https://help.github.com/articles/ignoring-files) for more information about ignoring files.|
|.rubocop.yml|This file contains the configuration for RuboCop.|
|.ruby-version|This file contains the default Ruby version.|
|.devcontainer/|This folder contains the Dev Container configuration|

### Opening the Blog Application in a Dev Container

> A Dev Container is a fully functional development environment running in a container

Reviewer: I suggested explaining what a Dev Container is at the start of the guide when we mention Dev Containers for the first time, so maybe if we decide to go that route, we can remove this introduction of what a Dev Container is?

on your machine. Our new Rails application comes with a Dev Container already
configured and ready to use.

We will use VSCode to spin up and work with our Dev Container. First we need to open
our app in VSCode. This can be done by opening VSCode and then opening our application's
folder, or directly from the command line within our application's directory with:

```bash
$ code .

> ```

Reviewer: I'm not sure that we need to explain _how_ to open a file in VSCode as it feels a bit too verbose, maybe we can just say to launch the VSCode application? I don't feel strongly about this if you feel that its better to go into the detail.

Reviewer 3: I think this is a good tip for new folks who don't yet know command line prompts. I say we keep it.

Author: I actually just realized the `code` command is not added to your path automatically when you install vscode. So I am going to remove this, other wise I will need to explain how to create it - which involves opening vscode 😅 so it's kind of a side quest at that point.

Author: But I tried to simplify it.

When we open the application, VSCode will recognize that a Dev Container is available

> and will prompt us to re-open the application in a Dev Container. Click the prompt

Reviewer: maybe we can allude to what the prompt is, something like "Folder contains a Dev Container configuration file. Reopen folder to develop in a container ([learn more](https://aka.ms/vscode-remote/docker))."

Reviewer: Maybe we should also add a reminder to have the Docker application running?

Author: I was thinking to avoid matching the language exactly as VScode could change it. But I will make it a bit more clear.

Reviewer: Sounds good!

and the Dev Container will begin initializing.

> Once the Dev Container setup is complete, VSCode will now be working with a fully
> setup development environment, with Ruby, Rails and all gems and dependencies installed.

Reviewer: Should this be?

```suggestion
Once the Dev Container setup is complete, your application will now be running in a fully
setup development environment; with Ruby, Rails, and all gems and dependencies installed.
```

Reviewer: What do you think about this rephrasing?

```suggestion
Once the Dev Container setup is complete, your development environment is ready to use, with Ruby, Rails, and all your dependencies installed. 
```

> You can open the terminal within VScode to verify:

Reviewer: ```suggestion
You can open the terminal within VSCode to verify that Rails is installed:
```

```bash
$ rails --version
Rails 7.2.0
```

> For the remainder of this guide, you will be working from the terminal in VSCode, which
> is your entry point for working your Application running inside the Dev Container.

Reviewer: ```suggestion
For the remainder of this guide, you will be working within VSCode which is your entry point for your Application running inside the Dev Container.
```

Reviewer: What do you think about? 

```suggestion
For the remainder of this guide, you will be working within VSCode which serves as your entry point for your application that will run in the Dev Container.
```
